pkg/cni/plugin.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2019 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cni
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"runtime"
+	"strings"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+
+	"github.com/containernetworking/plugins/pkg/ip"
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/containernetworking/plugins/pkg/utils/sysctl"
+)
+
+const (
+	// IPv4InterfaceArpProxySysctlTemplate allows proxy_arp on a given interface
+	IPv4InterfaceArpProxySysctlTemplate = "net.ipv4.conf.%s.proxy_arp"
+)
+
+// A NetConf structure represents a Multus network attachment definition configuration
+type NetConf struct {
+	types.NetConf
+	DeviceID string `json:"deviceID"`
+	MTU      int    `json:"mtu,omitempty"`
+}
+
+// EnvArgs structure represents inputs sent from each VMI via environment variables
+type EnvArgs struct {
+	types.CommonArgs
+	MAC types.UnmarshallableString `json:"mac,omitempty"`
+}
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func loadConf(bytes []byte) (*NetConf, string, error) {
```

Comment:
I'm not a Go expert, but I often heard that pointers in Go are overrated and it is better to just pass the value alone. Take a look at https://medium.com/@vCabbage/go-are-pointers-a-performance-optimization-a95840d3ef85for basic introduction. Unless we don't need a pointer for e.g. differentiation between nil and zero, let's try to use basic values. Feel free to argue with me here, I haven't figured out yet when to use them and when not, I'm just looking for a rule of thumb.

Reply:
I can live with that, no biggie.